### DISCUSSION TITLE
Getting NodeId by using name space uri

### DISCUSSION THREAD

**Discussion by bibhudatta-sahu90** [eclipse-milo/milo#946] [General]
Hello Everyone,
    I am using MILO SDK. In Opcua client, I would like to know how can I get the node id by using name space uri.
    
Thanks,
Bibhudatta Sahu

**Comment by kevinherron** [comment]
A namespace URI isn’t enough information to identify a Node or specify a NodeId.

**Reply by bibhudatta-sahu90** [reply]
Thank you Kevin. 

Let me describe the scenario, we subscribed few nodes on opcua server. After that we shutdown the client and close the application. When application restarts, opcua client again connected to server and re-subscribe already subscribed nodes (only those which we already subscribed before application restarts).

Lets user modify some configuration on server and add few namespaces to server address space, after that he restarts the server. Once the server restarts it will re-arrange the nodes and change its ids on server address space. 

For identifying the required node we are storing namespace uri (Or probably we can use urn or some related information) in some file. When application restarts we take those namespace uri information from file to identify nodes which we subscribed before closing the client connection. 

Overall, we are trying to achieve re-subscribe operation when application restarts. Could you please let me know what other information is required to identify or specify a node. And please guide us how can we achieve it.

Thanks,
Bibhudatta Sahu

**Reply by kevinherron** [reply]
Okay, I think I understand the issue. The namespaces can end up at different indices each time it restarts.

You can solve this by:
1. store absolute `ExpandedNodeId` instead of `NodeId`, which is similar to storing `(NamespaceUri + Identifier)` rather than `(NamespaceIndex + Identifier)`. We convert `NodeId` to an absolute `ExpandedNodeId` and then store a parseable version of that (see `ExpandedNodeId::absolute` and `ExpandedNodeId::toParseableString`).
2. when/after your client connects, read the current NamespaceTable (see `OpcUaClient::readNamespaceTable` for a helper method).
3. use the `NamespaceTable` to convert your stored `ExpandedNodeId`s into `NodeId` when it's time to re-subscribe everything (`ExpandedNodeId::toNodeId`).

Steps 2/3 really sort of happen all at the same time.

**Reply by bibhudatta-sahu90** [reply]
Thank you Kevin. I was following the steps. But I was not able to get the step 3. I would like to know how can I get ExpandedNodeId from the NamespaceTable and then converting into NodeId. Could you please provide some example for reference.

Thanks,
Bibhudatta Sahu

**Reply by kevinherron** [reply]
Your application has to store things in a format that will let you parse them into an ExpandedNodeId (see step 1).

Once you have one, then you just call `toNodeId` like I suggested:
```java
NamespaceTable namespaceTable = client.readNamespaceTable().get();

ExpandedNodeId xni = ...; // you somehow parsed this from stored information

NodeId nodeId = xni.toNodeId(namespaceTable).orElse(null); // might be null if the namespace URI no longer exists on the server
```

**Reply by bibhudatta-sahu90** [reply]
Thank you Kevin. It is helpful.
